### ISSUE TITLE
broken tests

### ISSUE THREAD

**Issue by ysangkok** [mit-dci/utreexo#91]
The current master 5333c202 is failing tests. I tried bisecting, and <s> I found that while 91fb4e9 runs `go test ./utreexo -v` without error, 710df49 ("fix one more", one commit after) fails (but not with a panic like master). But at this point I am not sure if it is the test or the implementation that is broken, so I don't know how to proceed.</s> I bisected wrong (confusing lack of panic with passing test), but it looks like it has been failing for a long time, so I will just give up on bisecting.

Failing tests:

* <s> TestRandTopDown (panic: got non-moving swap)</s>
* <s> TestForestAddDel</s>
* TestUndoRandom
* TestUndoFixed (this test is really small)

**Comment by ysangkok** [comment]
But the panic on master is "got a left empty here", which was only added after those two older commits I mentioned...

**Comment by ysangkok** [comment]
I tried swapping in removeV3 but I get the same error. Occasionally I get another error though:

```
panic: got non-moving swap [recovered]
	panic: got non-moving swap

goroutine 710386 [running]:
testing.tRunner.func1.1(0x554a40, 0x5bbe50)
	/usr/lib/go-1.14/src/testing/testing.go:941 +0x3d0
testing.tRunner.func1(0xc000124a20)
	/usr/lib/go-1.14/src/testing/testing.go:944 +0x3f9
panic(0x554a40, 0x5bbe50)
	/usr/lib/go-1.14/src/runtime/panic.go:967 +0x15d
github.com/mit-dci/utreexo/utreexo.(*Forest).swapNodes(0xc00022c060, 0x41d, 0x41d, 0x1, 0x11, 0xc00050cb00)
	/home/janus/go/src/github.com/mit-dci/utreexo/utreexo/forest.go:227 +0x539
github.com/mit-dci/utreexo/utreexo.(*Forest).removev4(0xc00022c060, 0xc0002a4000, 0x1d, 0x1d, 0x0, 0x0)
	/home/janus/go/src/github.com/mit-dci/utreexo/utreexo/forest.go:137 +0x523
github.com/mit-dci/utreexo/utreexo.TestRandTopDown(0xc000124a20)
	/home/janus/go/src/github.com/mit-dci/utreexo/utreexo/transform_test.go:119 +0x325
```

I wonder if removeV3 was maintaining the required invariants when it was removed.

**Comment by ysangkok** [comment]
Ah, TestRandTopDown was still using removeV4 when that panic occurred. Using removeV3 it seems to run fine.

**Comment by ysangkok** [comment]
TestUndoFixed seems to fail independently of using removeV3 or removeV4.

**Comment by ysangkok** [comment]
Ok, I don't know why I wrote in the initial issue that 91fb4e9 runs without error because running it now three times, it failed every time.

**Comment by ysangkok** [comment]
It seems like TestUndoFixed asserts that the tree tops are identical after undoing. I am wondering if the undo is implemented in a way such that that would be the case, especially since the Undo implementation contains two TODOs...

**Comment by adiabat** [comment]
Fixed in #101 

The actual fix was simple enough once I narrowed it down.  
I'll try to keep green checkmarks now!
